src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -362,6 +362,41 @@ static void uct_ib_mlx5_devx_mr_lru_cleanup(uct_ib_mlx5_md_t *md)
     kh_destroy_inplace(rkeys, &md->lru_rkeys.hash);
 }
 
+static ucs_status_t
+uct_ib_mlx5_devx_reg_ksm(uct_ib_mlx5_md_t *md, uct_ib_mlx5_devx_mem_t *memh,
+                         uct_ib_mr_type_t mr_type, uint32_t iova_offset,
+                         int atomic, uint32_t mkey_index, const char *reason,
+                         struct mlx5dv_devx_obj **mr_p, uint32_t *mkey)
+{
+    uct_ib_mlx5_devx_mr_t *mr = &memh->mrs[mr_type];
+    uint64_t iova             = (uint64_t)memh->address + iova_offset;
+    void *address             = uct_ib_mlx5_devx_memh_base_address(memh);
+    ucs_status_t status;
+    size_t length;
+
+    if (memh->super.flags & UCT_IB_MEM_MULTITHREADED) {
+        ucs_assertv_always(mr->ksm_data->length > iova_offset,
+                           "mr->ksm_data->length=%ld iova_offset=%u",
+                           mr->ksm_data->length, iova_offset);
+        length = mr->ksm_data->length - iova_offset;
+        status = uct_ib_mlx5_devx_reg_ksm_mt(md, atomic, address, mr->ksm_data,
+                                             length, iova, mkey_index, reason,
+                                             mr_p, mkey);
+    } else {
+        length = mr->super.ib->length;
+        status = uct_ib_mlx5_devx_reg_ksm_contig(md, mr, address, iova, atomic,
+                                                 mkey_index, reason, mr_p, mkey);
+    }
+
+    if (status == UCS_OK) {
+        ucs_debug("KSM registered memory %p..%p offset 0x%x%s on %s rkey 0x%x",
+                  address, UCS_PTR_BYTE_OFFSET(address, length),
+                  iova_offset, atomic ? " atomic" : "",
+                  uct_ib_device_name(&md->super.dev), *mkey);
```

Comment:
maybe put the debug message inside uct_ib_mlx5_devx_reg_ksm_mt and uct_ib_mlx5_devx_reg_ksm_contig, to distinguish between them in the log?